javalib/src/main/scala/java/nio/channels/FileChannel.scala

```diff
@@ -113,24 +124,30 @@ object FileChannel {
       mode.append("s")
     }
 
-    val file = path.toFile()
-    val raf = new RandomAccessFile(file, mode.toString)
+    val raf = tryRandomAccessFile(path.toString, mode.toString)
 
-    if (writing && options.contains(TRUNCATE_EXISTING)) {
-      raf.setLength(0L)
-    }
+    try {
+      if (writing && options.contains(TRUNCATE_EXISTING)) {
+        raf.setLength(0L)
+      }
 
-    if (writing && options.contains(APPEND)) {
-      raf.seek(raf.length())
-    }
+      if (writing && options.contains(APPEND)) {
+        raf.seek(raf.length())
+      }
 
-    new FileChannelImpl(
-      raf.getFD(),
-      Some(file),
-      deleteFileOnClose = options.contains(StandardOpenOption.DELETE_ON_CLOSE),
-      openForReading = true,
-      openForWriting = writing
-    )
+      new FileChannelImpl(
+        raf.getFD(),
+        Some(path.toFile()),
+        deleteFileOnClose =
+          options.contains(StandardOpenOption.DELETE_ON_CLOSE),
+        openForReading = true,
+        openForWriting = writing
+      )
+    } catch {
+      case e: Throwable =>
+        raf.close()
+        throw e
+    }
```

Comment:
I think you should probably `close()` in a `finally` block unless I am missing something. Either way I think you need to swallow the exception from `close()`. Unless this is the callers obligation to catch the exception and `close()`. I think if the exception is thrown and nothing reasonably can be done then attempting to `close()` the resource is ok but not 100% sure that is the APIs responsibility.

```scala
finally {
  try {
    if (raf != null) {
        raf.close();
   } catch { case _: Throwable => } // I think this is the way to do it
}
```

Reply:
Like a `closeAndSwallow(c: Closeable)` utility method?

Reply:
Eric,  Thank you for your review & your suggestions.
The code you mentioned is at the end of 
```
 def open(
      path: Path,
      options: Set[_ <: OpenOption],
      attrs: Array[FileAttribute[_]]
  ): FileChannel 
```
On success, it returns an __open__ FileChannel.  That means that the underlying 
`raf` should only be closed on error. If should be closed so that we do not lose
OS file descriptors (fd's).  

Putting the close in `finally` block would pass back
a `FileChannel` with a closed underlying file descriptor.  My first mental drafts
went that way before I went "Oops!".

RandomAccessFile#close is documented as possibly returning an IOException.
That should be exceedingly rare, but if it _does_ happen, somebody almost
certainly wants to know about it.      

I have seen code in other places where a (unix) unconditional close(fd) is done
and the return value ignored.  This is usually in (poor) code where the fd may or
may not be open to begin with (or may be broken underneath).

In the code you identified, the `try` will only be entered if the RandomAccessFile
is open. The strongly presumed state in the `catch` block is that the file is open.
Again, any error is probably something someone wants to know about.  Something
in that World is not right (and probably leaking fds all over the place).

`FileChannel#open` is documented in Java 8 docs as `throws IOException`.
By my understanding of the contract, it is the responsibility of the caller to
decide what to do with an IOException and the responsibility of the 
executing method to reliably bubble up any Exceptions encountered.

Do you concur?  

Can you suggest a minimal comment or two that I could put in the code 
to capture/explain the intent?  The next person in the code is likely to 
be either of us, and I am not sure I could trace this again. (notice no emoticon)

Reply:
Line 148 is unsafe (no null check and calling `close()` without a `catch`) and should be similar to the `try` `catch` I provided without the `finally`. That is why I mentioned my hesitation about the `finally` because I was unsure without looking at the bigger picture.

Reply:
I really shouldn't be trying to review this if I can't dedicate enough time to do it properly - sorry.